### PR TITLE
James 1925 Big refacoring of Mailboxes

### PR DESCRIPTION
**Pull request by chibenwa**
First few commits are needed to do clean code.

Then adapt ACL, introduce DAO , modify MailboxMapper, solve some bugs, and add integration tests about original issue.

### CODE REVIEW

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxDAOTest.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        List<SimpleMailbox> mailboxes = testee.retrieveAllMailboxes()
+            .join()
+            .collect(Guavate.toImmutableList());
+
+        assertThat(mailboxes).containsOnly(mailbox1);
+    }
+
+    @Test
+    public void retrieveAllMailboxesShouldReturnMultiMailbox() {
```
Comment (quynhn) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxDAOTest.java]: s/Mailbox/Mailboxes/

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/MailboxBaseTupleUtil.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.mail.utils;
+
+import org.apache.james.backends.cassandra.init.CassandraTypesProvider;
+import org.apache.james.mailbox.cassandra.table.CassandraMailboxTable;
+
+import com.datastax.driver.core.UDTValue;
+
+public class MailboxBaseTupleUtil {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/MailboxBaseTupleUtil.java]: MailboxUDTs ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxPathDAO.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    public CompletableFuture<Stream<CassandraIdAndPath>> listUserMailboxes(String namespace, String user) {
+        return cassandraAsyncExecutor.execute(
+            selectAll.bind()
+                .setUDTValue(NAMESPACE_AND_USER, mailboxBaseTupleUtil.createMailboxBaseUDT(namespace, user)))
+            .thenApply(resultSet -> CassandraUtils.convertToStream(resultSet).map(this::fromRowToCassandraIdAndPath));
+    }
+
+    private CassandraIdAndPath fromRowToCassandraIdAndPath(Row row) {
+        return new CassandraIdAndPath(
+            CassandraId.of(row.getUUID(CassandraMessageIds.MAILBOX_ID)),
```
Comment (quynhn) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxPathDAO.java]: My fault, it should be "MAILBOX_ID"

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 4]
```diff
@@ -121,10 +121,15 @@ public Mailbox findMailboxById(MailboxId id) throws MailboxException {
...
         SimpleMailbox cassandraMailbox = (SimpleMailbox) mailbox;
-        if (cassandraMailbox.getMailboxId() == null) {
-            cassandraMailbox.setMailboxId(CassandraId.timeBased());
-        }
-        boolean applied = mailboxPathDAO.save(mailbox.generateAssociatedPath(), (CassandraId) cassandraMailbox.getMailboxId())
+
+        CassandraId cassandraId = retrieveId(cassandraMailbox);
+        cassandraMailbox.setMailboxId(cassandraId);
+
+        boolean applied = mailboxDAO.retrieveMailbox(cassandraId)
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: what does it solve, there's not test change here ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 5]
```diff
@@ -135,23 +124,32 @@ public void save(Mailbox mailbox) throws MailboxException {
...
+            throw new MailboxExistsException(mailbox.generateAssociatedPath().asString());
+        }
     }
 
     @Override
     public boolean hasChildren(Mailbox mailbox, char delimiter) {
-        final Pattern regex = Pattern.compile(Pattern.quote( mailbox.getName() + String.valueOf(delimiter)) + ".*");
-        return getMailboxFilteredByNamespaceAndUserStream(mailbox.getNamespace(), mailbox.getUser())
-            .anyMatch((row) -> regex.matcher(row.getString(NAME)).matches());
+        Pattern regex = Pattern.compile(Pattern.quote( mailbox.getName() + String.valueOf(delimiter)) + ".*");
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: startswith is not enough ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: It's catched by MPT

File: server/container/util-java8/src/main/java/org/apache/james/util/CompletableFutureUtil.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+public class CompletableFutureUtil {
+
+    public static <T> CompletableFuture<Stream<T>> allOf(Stream<CompletableFuture<T>> futureStream) {
+        return futureStream
+            .map(future -> future.thenApply(Stream::of))
+            .reduce((future1, future2) ->
+            future1.thenCompose(
```
Comment (rouazana) [inline server/container/util-java8/src/main/java/org/apache/james/util/CompletableFutureUtil.java]: indent

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java [hunk 7]
```diff
@@ -66,16 +66,16 @@
...
     }
 
-    public CassandraACLMapper(CassandraId cassandraId, Session session, int maxRetry, CodeInjector codeInjector) {
+    public CassandraACLMapper(CassandraId cassandraId, Session session, CassandraAsyncExecutor cassandraAsyncExecutor, int maxRetry, CodeInjector codeInjector) {
         Preconditions.checkArgument(maxRetry > 0);
         Preconditions.checkArgument(cassandraId != null);
         this.cassandraId = cassandraId;
         this.session = session;
-        this.executor = new CassandraAsyncExecutor(session);
+        this.executor = cassandraAsyncExecutor;
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java]: the executor is already initialized with the good session?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperConcurrencyTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    @Test
+    public void saveShouldBeThreadSafe() throws Exception {
+        boolean termination = new ConcurrentTestRunner(THREAD_COUNT, OPERATION_COUNT,
+            (a, b) -> testee.save(new SimpleMailbox(MAILBOX_PATH, UID_VALIDITY)))
+            .run()
+            .awaitTermination(1, TimeUnit.MINUTES);
+
+        assertThat(termination).isTrue();
+        assertThat(testee.list()).hasSize(1);
+    }
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapperConcurrencyTest.java]: no concurrent test on renaming?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java (hunk 7 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java]: Yes. There's a flow with ACLMapper  as mailboxId is a field. Then we need to create it per request. I did not wanted some weird threadpool issues.

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxDAOTest.java (hunk 1 above)
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxDAOTest.java]: Good catch

File: server/container/util-java8/src/main/java/org/apache/james/util/CompletableFutureUtil.java (hunk 6 above)
Comment (chibenwa) [inline server/container/util-java8/src/main/java/org/apache/james/util/CompletableFutureUtil.java]: I just had some parsing troubble in my IDE. I added variable extraction as a workaround.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java (hunk 4 above)
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: so it belongs to another commit ?

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: Yes commit name is misleading. Should be fixuo

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: Yes commit name is misleading. Should be fixuo

**Comment by chibenwa** [comment]
Thanks @strokyl ;-)

**Comment by chibenwa** [comment]
test this please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 9]
```diff
@@ -52,103 +41,147 @@
...
         Preconditions.checkArgument(mailbox instanceof SimpleMailbox);
         SimpleMailbox cassandraMailbox = (SimpleMailbox) mailbox;
+
+        CassandraId cassandraId = retrieveId(cassandraMailbox);
+        cassandraMailbox.setMailboxId(cassandraId);
+        if (mailbox.getName().length() >= MAX_SIZE_FOR_MAILBOX_NAME) {
+            throw new TooLongMailboxNameException("too long mailbox name");
+        }
+        try {
+            boolean applied = mailboxDAO.retrieveMailbox(cassandraId)
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: 1. Read mailboxDAO to retrieve potential old path
    2. (thenCompose) Inset new Path in mailboxPathDAO
    3. thenCompose if success, parallel(delete old path with mailboxPathDAO, save in mailboxDAO)

Add a test on moving mailbox. If new name too long, old mailboxPath should still exists

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 10]
```diff
@@ -57,6 +57,8 @@
 
     public static final String CLUSTERING_COLUMNS_IS_TOO_LONG = "The sum of all clustering columns is too long (65544 > 65535)";
 
+    public static final int MAX_SIZE_FOR_MAILBOX_NAME = 64* (1 << 10);
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: 1 << 10 is 2^10 = 1 ko

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java [hunk 11]
```diff
@@ -129,21 +128,20 @@ public void save(Mailbox mailbox) throws MailboxException {
...
-        if (mailbox.getName().length() >= MAX_SIZE_FOR_MAILBOX_NAME) {
-            throw new TooLongMailboxNameException("too long mailbox name");
-        }
         try {
-            boolean applied = mailboxDAO.retrieveMailbox(cassandraId)
-                    .thenCompose(optional -> optional
-                            .map(storedMailbox -> mailboxPathDAO.delete(storedMailbox.generateAssociatedPath()))
-                            .orElse(CompletableFuture.completedFuture(null)))
-                    .thenCompose(any -> mailboxPathDAO.save(mailbox.generateAssociatedPath(), cassandraId))
+            boolean applied = mailboxPathDAO.save(mailbox.generateAssociatedPath(), cassandraId)
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: Please run SMTP test for cassandra, I guess you are breaking renames

Step 1 : retrieve old path (it's not attached to the mailbox, see rename in StoreMailboxManager)
Step 2 : Try insert on mailboxPath table for new entry
Step 3 : Then update Mailbox table and remove old entry in mailboxPath table

Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: CassandraMailboxTest is running fine. Do you expect something other?

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: Hmm I was expecting it to fail on mailboxRename

I must be paranoïd. 

If green, then should be fine.

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: I can 't approuve my own pull request but 👍 for your changes

You inverted some future steps compared to what I proposed, however result is the same. I was just confused.

Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxMapper.java]: we could had some parrellism in a future PR

**Comment by aduprat** [comment]
merged
